specs/GetSecurityDescriptorForAppContainerNames/GetSecurityDescriptorForAppContainerNames.md

```diff
@@ -0,0 +1,87 @@
+# Easing and streamlining named object sharing between packaged and Win32
+
+This feature is a flat C API that takes a list of Package Family Names (PFNs) and access masks, and returns a security descriptor
+
+# Background
+
+Named objects (mutexes, events, waitable timers) must be ACLed during creation in order to be shared with packaged applications.
+
+Currently, [our guidance](https://docs.microsoft.com/en-us/windows/uwp/communication/sharing-named-objects) for this scenario is to write a large amount of complex, cumbersome, and largely boilerplate code. ([See here for a taste](https://docs.microsoft.com/en-us/windows/win32/api/securityappcontainer/nf-securityappcontainer-getappcontainernamedobjectpath#examples)).
+
+If we are to meet developers where they are, we need to reduce the friction of Win32 and UWP interop, and make it as easy as possible to share named objects.
+
+# Description
+
+Add a flat C API that takes an array of PFNs and access masks, and produces the needed `SECURITY_ATTRIBUTES` object.
+
+# Examples
+
+```c++
+#include <wil/resource.h>
+
+wil::unique_event CreateShareableEvent(PCWSTR name)
+{
+    wil::unique_hlocal_security_descriptor sd;
+    AppContainerAccess access[1] =
+      {{L"Contoso.Test.App_12345678",
+        EVENT_MODIFY_STATE | SYNCHRONIZE}};
+
+    THROW_IF_FAILED(GetSecurityDescriptorForAppContainerNames(
+        1, access, EVENT_MODIFY_STATE | SYNCHRONIZE, &sd);
+
+    SECURITY_ATTRIBUTES sa;
+    sa.nLength = sizeof(sa);
+    sa.lpSecurityDescriptor = sd.get();
+    sa.bInheritHandle = FALSE;
+
+    wil::unique_event result;
+    result.create(wil::EventOptions::None, name, &sa);
+    // aka, ::CreateEvent(&sa, FALSE, FALSE, name);
+    return result;
+}
+```
+
+# API Details
+
+```c
+struct AppContainerAccess
+{
+    PCWSTR appContainerName;
+    uint32_t accessMask;
+};
+
+STDAPI GetSecurityDescriptorForAppContainerNames(
+    uint32_t countOfAppContainerNames,
+    _In_reads_(countOfAppContainerNames)
+        const AppContainerAccess* appAccess,
+    uint32_t userAccessMask,
+        _Outptr_ PSECURITY_DESCRIPTOR* securityDescriptor
+)
+```
+
+If the function succeds, the returned `SECURITY_DESCRIPTOR` must be freed by calling [LocalFree](https://docs.microsoft.com/en-us/windows/desktop/api/winbase/nf-winbase-localfree).
+
+# API Notes
+
+Some alternative designs were considered.
+- Single method, takes a list of permissions corresponding to the list of PFNs. (Current choice)
+  - Pros
+    - Handles both complicated case and simple case in a single call.
+    - API is straightforward, no extra/optional/multimode parameters.
+    - Only a single method to search/remember/code.
+  - Cons
+    - Simple case is not "optimally" simple - requires filling array with a single value.
```

Comment:
Sorta-simple version:

```cpp
THROW_IF_FAILED(GetSecurityDescriptorForAppContainerNames(
    1, std::array<AppContainerAccess, 1>{ { L"Contoso.Test.App_12345678", EVENT_MODIFY_STATE | SYNCHRONIZE } }.data(),
    EVENT_MODIFY_STATE | SYNCHRONIZE,
   &sd));
```

However you still wouldn't want to do this because it creates a temporary at runtime.

```cpp
```cpp
THROW_IF_FAILED(GetSecurityDescriptorForAppContainerNames(
    1, std::array<AppContainerAccess, 1>{ { L"Contoso.Test.App_12345678", EVENT_MODIFY_STATE | SYNCHRONIZE } }.data(),
    EVENT_MODIFY_STATE | SYNCHRONIZE,
   &sd));
```

This does build a temporary at runtime, though.

```cpp
static constexpr AppContainerAccess access{ L"Contoso.Test.App_12345678", EVENT_MODIFY_STATE | SYNCHRONIZE };

THROW_IF_FAILED(GetSecurityDescriptorForAppContainerNames(
    1, &access,
    EVENT_MODIFY_STATE | SYNCHRONIZE,
   &sd));
```

Reply:
DISCUSSION: Neato!